Lib/pydoc.py

```diff
@@ -1617,13 +1617,11 @@ def pipepager(text, cmd):
 def tempfilepager(text, cmd):
     """Page through text by invoking a program on a temporary file."""
     import tempfile
-    filename = tempfile.mktemp()
-    with open(filename, 'w', errors='backslashreplace') as file:
-        file.write(text)
-    try:
+    with tempfile.TemporaryDirectory() as tempdir:
+        filename = os.path.join(tempdir, 'pydoc.out')
+        with open(filename, 'w', errors='backslashreplace') as file:
```

Comment:
In Windows, the file should be opened using `encoding=os.device_encoding(0)`, at least until a builtin "conin" encoding is supported.

"more.com" decodes input text using the console's input code page, which defaults to the system OEM code page. The above `open` uses the default locale encoding (not the current locale encoding, unlike POSIX). The default locale encoding is the active code page of the current process, which is the system ANSI code page unless overridden to UTF-8 in the application manifest. Due to the mismatched encodings, "more.com" will display non-ASCII values as mojibake, particularly in a Western locale for which the system ANSI and OEM code pages are different. For example, "ß" will be displayed as "▀" if it's encoded with code page 1252 and decoded with code page 437.

Reply:
This really isn't my area (and I have a lot of Googling to do after your comment!) but I want to check that the `fd` should be 0 rather than `sys.stdout.fileno()`? Also, does anyone know if adding this kwarg would have a side-effect on other platforms (i.e. should we avoid using it unless on Windows)?

Reply:
I only intended for this to apply to Windows. For example:

```python
encoding = os.device_encoding(0) if sys.platform == 'win32' else None
```
>  fd should be 0 rather than `sys.stdout.fileno()`

`os.device_encoding(fd)` returns `None` in Windows, except for file descriptors 0-2 if `isatty(fd)` is true. The latter is true for any character device, including CON, CONIN$, CONOUT$, NUL, COM\<N\>, and LPT\<N\>. For fd 0, `device_encoding` returns the console input code page from `GetConsoleCP()`, and for fds 1-2 it returns the console output code page from `GetConsoleOutputCP()`. I have an open issue to provide a real implementation, instead of hard-coded assumptions, that gets the OS file handle from the file descriptor and checks whether it's a console input or output handle to determine which console code page to return.

The suggestion for now is to use fd 0, which is hard coded to return the console input code page if the fd is a character device and the process is attached to a console session. Otherwise the encoding will be `None`, and the file will use the locale default encoding. If you want something more reliable, you could use ctypes:

```python
import ctypes as _ctypes
_kernel32 = _ctypes.WinDLL('kernel32', use_last_error=True)

def _get_console_cp():
    cp = _kernel32.GetConsoleCP()
    if cp == 0:
        cp = _kernel32.GetACP()
    if cp == 65001:
        return 'UTF-8'
    return f'cp{cp}'
```

Reply:
Thanks Eryk for the explanation. I decided to go for just `os.device_encoding(0)` for simplicity.